src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -587,22 +503,32 @@ internal static string GetServerDefinedTypeFullName(Type type)
         /// <returns>Client clr name.</returns>
         internal static string GetClientFieldName(Type t, string serverDefinedName)
         {
+            ODataTypeInfo typeInfo = GetODataTypeInfo(t);
+                        
             List<string> serverDefinedNames = serverDefinedName.Split(',').Select(name => name.Trim()).ToList();
             List<string> clientMemberNames = new List<string>();
             foreach (var serverSideName in serverDefinedNames)
             {
-                FieldInfo memberInfo = t.GetField(serverSideName) ?? t.GetFields().ToList().Where(m =>
-                {
-                    OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)m.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
-                    return originalNameAttribute != null && originalNameAttribute.OriginalName == serverSideName;
-                }).SingleOrDefault();
+                string memberInfoName;
 
-                if (memberInfo == null)
+                if (!typeInfo.ServerSideNameDict.TryGetValue(serverSideName, out memberInfoName))
                 {
-                    throw c.Error.InvalidOperation(c.Strings.ClientType_MissingProperty(t.ToString(), serverSideName));
+                    FieldInfo memberInfo = t.GetField(serverSideName) ?? t.GetFields().ToList().Where(m =>
+                    {
+                        OriginalNameAttribute originalNameAttribute = (OriginalNameAttribute)m.GetCustomAttributes(typeof(OriginalNameAttribute), false).SingleOrDefault();
+                        return originalNameAttribute != null && originalNameAttribute.OriginalName == serverSideName;
+                    }).SingleOrDefault();
+
+                    if (memberInfo == null)
+                    {
+                        throw c.Error.InvalidOperation(c.Strings.ClientType_MissingProperty(t.ToString(), serverSideName));
+                    }
+
+                    memberInfoName = memberInfo.Name;
+                    typeInfo.ServerSideNameDict[serverSideName] = memberInfoName;
                 }
 
-                clientMemberNames.Add(memberInfo.Name);
+                clientMemberNames.Add(memberInfoName);
             }
 

```

Comment:
implement a method "GetClientFieldName" in ODataTypeInfo that takes a type and a single server defined name and returns the client name.  Keep the logic for splitting the string, calling the GetClientFieldName method, and joining the results in this method.